implement compare for ion-js-cli test-driver

desaikd opened the pull request:
**Description:**

As part of ion-test-driver, this commit is for ion-js-cli compare. It supports:

 - --output(-o) : rewrite given input files to specified output format
 - --output-fomrat(-f): specify ion output format
 - --error-report(-e): generate error report to specified location 
- --comparison-type(-y): comparison semantics to be used with compare command

**Completeted Tasks:**

1. Implement comparisonContext
2. Implement comprisonReport
3. comparison using event stream equals method
4. currently comparisonReport prints only basic comparison for SCALAR events.

**TO DO:**

- [ ] Support comparisonReport for basic comparison of LIST, SEXP and STRUCT event type.
- [ ] Support equivs and non_equivs comparison 

**Test:**

tested using multiple input files for comparison (binary, text, pretty, events).
e.g. ts-node compare example1.ion example2.ion --output comparison_report

desaikd committed:
implement compare for ion-js-cli test-driver

tgregg commented on src/Cli.ts:
```diff
@@ -75,6 +90,41 @@ export class IonCliError {
+        writer.stepIn(IonTypes.STRUCT);
+        writer.writeFieldName('result');
+        writer.writeSymbol(this.result);
+        this.lhs.writeComparisonContext(writer, true, event_index);
+        this.rhs.writeComparisonContext(writer, false, event_index);
```
I recommend just setting the field names here instead of in `writeComparisonContext`.

desaikd replied:
Should I move `writeComparisonContext` method here in `IonComparisonReport`? So that for `lhs` and `rhs` I won't have to write the same lines of code twice.

tgregg replied:
No, I think `writeComparisonContext` is fine as a method of `ComparisonContext`. Which lines of code are you referring to that you'd have to repeat?

desaikd replied:
I think I misunderstood what you mean by setting field names here. I got it now. will change that. Thanks.

tgregg commented on src/Compare.ts:
```diff
@@ -0,0 +1,134 @@
+    'comparison-type': {
+        alias: "y",
+        default: 'basic',
+        choices: ['basic', 'equivs', 'non-equivs', 'equivs_timeline'],
```
equivs_timeline -> equiv-timeline

tgregg commented on src/Cli.ts:
```diff
@@ -75,6 +90,41 @@ export class IonCliError {
+        this.message = message;
+    }
+
+    writeErrorReport(comparisonReportFile: WriteStream| NodeJS.WriteStream, event_index) {
```
Would this be more accurately named `writeComparisonReport`?

tgregg commented on src/Compare.ts:
```diff
@@ -0,0 +1,134 @@
+        ionOutputWriter.writeFieldName("event");
+        this.getEventStream().getEvents()[event_index].write(ionOutputWriter);
+        ionOutputWriter.stepOut();
+        ionOutputWriter.writeFieldName("event_index");
+        ionOutputWriter.writeInt(event_index);
```
The event_index field is part of the ComparisonContext struct, so these lines should be moved before the stepOut().

tgregg commented on src/Compare.ts:
```diff
@@ -0,0 +1,134 @@
+    compareFiles(ionOutputWriter: Writer, args: IonCompareArgs): void {
+        for (let pathFirst of args.getInputFiles()) {
+            for(let pathSecond of args.getInputFiles()) {
+                if(pathFirst === pathSecond) {
```
A path will still need to be compared against itself for equivs, non-equivs, and equiv-timeline semantics.

desaikd committed:
added support for basic comparison of list and struct

desaikd commented:
**Commit:**  d7b63de

**Tasks:**

1. Implemented `ComparisonResult` in `Cli` which stores the message and event index for a particular comparison. 
2. Instead of using `equals` method implemented `compare` in `IonEvent` which compares the events and creates a `ComparisonResult`. (Changed `equals` to use compare to return boolean)
3. Added support for basic comparison of List and Struct.

**Test:**
Tested with different struct and list values for comparison. 

1. Struct comparison when not equal will return the `CONTIANER_START` event along with a message representing the error related to it.
2. List comparison when not equal will return the particular `SCALAR` event along with a message of their differing Ion Values.

tgregg commented on src/Cli.ts:
```diff
@@ -97,31 +97,44 @@ export enum ComparisonResultType {
+export class ComparisonResult {
+    message: string;
     result: ComparisonResultType;
+    actualIndex: number;
+    expectedIndex: number;
```
Out of curiosity, why are you choosing to store the indices rather than the ComparisonContexts?

desaikd replied:
Because otherwise I have to pass the entire `ComparisonContext` object to `IonEvent` which is a single event from a `ComparisonContext`'s event stream and there's no way to get the particular event from event stream except through index. So I thought to just pass the index rather than passing the object. Does this answer your concern?

tgregg commented on src/IonEvent.ts:
```diff
@@ -482,20 +538,20 @@ class IonSexpEvent extends AbsIonContainerEvent {
             let child = container[i];
-            if (!child.equals(expectedContainer[i])) {
-                return false;
+            if (child.compare(expectedContainer[i]).result == ComparisonResultType.NOT_EQUAL) {
+                return child.compare(expectedContainer[i]);
```
Looks like this compare is performed twice. Consider storing the result and returning it later.

tgregg commented on src/IonEvent.ts:
```diff
@@ -459,20 +514,21 @@ class IonListEvent extends AbsIonContainerEvent {
-                return false;
+            if (child.compare(expectedContainer[i]).result == ComparisonResultType.NOT_EQUAL) {
+                return new ComparisonResult(ComparisonResultType.NOT_EQUAL, child.ionValue + " vs. " +
+                    expectedContainer[i].ionValue, i + 1, i + 1);
```
Why +1 ?

desaikd replied:
Considering the first event would be container start (`LIST`) and then I am rotating over its child events.

tgregg reviewed (approved):
Just minor comments. Feel free to complete this PR and continue in a new one.

desaikd committed:
modified as the changes requested

desaikd committed:
Implement equivs, equiv-timeline and non-equivs comparison

* Implement equivs and non-equivs comparison

* implemet equiv-timeline comparison

* changes to equiv-timeline comparison
